apollo-router/src/plugins/router_limits/mod.rs

```diff
@@ -0,0 +1,366 @@
+            .get_limits()
+            .ok_or("License limits found during plugin initialization but failed to get limits in constructor phase of router_limits")
+            ?;
+
+        let tps = limits
+            .tps
+            .ok_or("License limits defined but no TPS claim defined")?;
+
+        let capacity = NonZeroU64::new(tps.capacity as u64)
+            .ok_or("Failed to convert TPS capacity into a usable value")?;
+
+        Ok(Self {
+            tps: TpsLimitConf {
+                capacity,
+                interval: tps.interval,
+            },
+        })
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        ServiceBuilder::new()
+            .map_future_with_request_data(
+                |req: &router::Request| req.context.clone(),
+                move |ctx, future| async {
+                    let response: Result<RouterResponse, BoxError> = future.await;
+                    match response {
+                        Ok(ok) => Ok(ok),
+                        Err(err) if err.is::<Overloaded>() => {
+                            u64_counter!(
+                                "apollo.router_limits.tps.limit_enforced",
+                                "TPS Rate Limiting triggered",
+                                1u64
+                            );
+                            let error = graphql::Error::builder()
+                                .message("Your request has been rate limited")
+                                .extension_code("ROUTER_TPS_LIMIT_REACHED")
+                                .build();
+                            Ok(RouterResponse::error_builder()
+                                .status_code(StatusCode::TOO_MANY_REQUESTS)
+                                .error(error)
+                                .context(ctx)
+                                .build()
+                                .expect("should build overloaded response"))
+                        }
+                        Err(err) => Err(err),
+                    }
+                },
+            )
+            .load_shed()
+            .layer(RateLimitLayer::new(
+                self.tps.capacity.into(),
+                self.tps.interval,
+            ))
+            .service(service)
+            .boxed()
+    }
+}
+
+register_private_plugin!("apollo", "router_limits", RouterLimits);
+
+#[cfg(test)]
+mod test {
+    use serde_json::json;
+    use tower::Service;
+
+    use super::*;
+    use crate::metrics::FutureMetricsExt;
+    use crate::plugin::test::MockRouterService;
+    use crate::plugin::DynPlugin;
+    use crate::plugins::test::PluginTestHarness;
+    use crate::services::RouterRequest;
+    use crate::uplink::license_enforcement::LicenseLimits;
+    use crate::uplink::license_enforcement::LicenseState;
+    use crate::uplink::license_enforcement::TpsLimit;
+
+    const APOLLO_ROUTER_LIMITS: &str = "apollo.router_limits";
+
+    async fn get_router_limits_plugin(
```

Comment:
Why no use the test harness for all tests?

Reply:
fixed